packages/signature-controller/src/SignatureController.ts

```diff
@@ -262,38 +254,64 @@ export class SignatureController extends BaseControllerV2<
     this.update(() => getDefaultState());
   }
 
+  /**
+   * Reject all unapproved messages of any type.
+   *
+   * @param reason - A message to indicate why.
+   */
+  rejectUnapproved(reason?: string) {
+    this.#rejectUnapproved(this.#messageManager, reason);
+    this.#rejectUnapproved(this.#personalMessageManager, reason);
+    this.#rejectUnapproved(this.#typedMessageManager, reason);
+  }
+
+  /**
+   * Clears all unapproved messages from memory.
+   */
+  clearUnapproved() {
+    this.#clearUnapproved(this.#messageManager);
+    this.#clearUnapproved(this.#personalMessageManager);
+    this.#clearUnapproved(this.#typedMessageManager);
+  }
+
   /**
    * Called when a Dapp uses the eth_sign method, to request user approval.
    * eth_sign is a pure signature of arbitrary data. It is on a deprecation
    * path, since this data can be a transaction, or can leak private key
    * information.
    *
-   * @param msgParams - The params passed to eth_sign.
+   * @param messageParams - The params passed to eth_sign.
    * @param [req] - The original request, containing the origin.
    * @returns Promise resolving to the raw data of the signature request.
    */
   async newUnsignedMessage(
-    msgParams: MessageParams,
+    messageParams: MessageParams,
     req: OriginalRequest,
   ): Promise<string> {
-    if (!this.#isEthSignEnabled()) {
-      throw ethErrors.rpc.methodNotFound(
-        'eth_sign has been disabled. You must enable it in the advanced settings',
-      );
-    }
-
-    const data = this.#normalizeMsgData(msgParams.data);
-
-    // 64 hex + "0x" at the beginning
-    // This is needed because Ethereum's EcSign works only on 32 byte numbers
-    // For 67 length see: https://github.com/MetaMask/metamask-extension/pull/12679/files#r749479607
-    if (data.length !== 66 && data.length !== 67) {
-      throw ethErrors.rpc.invalidParams(
-        'eth_sign requires 32 byte message hash',
-      );
-    }
-
-    return this.#messageManager.addUnapprovedMessageAsync(msgParams, req);
+    return this.#newUnsignedAbstractMessage(
+      this.#messageManager,
+      ApprovalType.EthSign,
+      'Message',
+      this.#signMessage.bind(this),
+      messageParams,
+      req,
+      (params) => {
```

Comment:
Would it be cleaner to create separate private methods for the validation logic, that you then `bind` into these callback arguments, like we have for the `signMessage` callback?

Reply:
Done